File: src/main/kotlin/club/sk1er/elementa/font/VanillaFontRenderer.java
package club.sk1er.elementa.font;

import club.sk1er.mods.core.universal.UGraphics;
import club.sk1er.mods.core.universal.UMinecraft;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;

public class VanillaFontRenderer implements FontProvider {
    @Override
    public float getStringWidth(@NotNull String string, float pointSize) {
        return UMinecraft.getFontRenderer().getStringWidth(string);
    }

    @Override
    public void drawString(@NotNull String string, @NotNull Color color, float x, float y, float originalPointSize, boolean shadow, @Nullable Color shadowColor) {
        if (shadowColor == null) {
            UGraphics.drawString(string, x, y, color.getRGB(), shadow);
        } else {
            UGraphics.drawString(string, x, y, color.getRGB(), shadowColor.getRGB());
        }
    }
}

File: src/main/kotlin/club/sk1er/elementa/VanillaFontRenderer.kt
package club.sk1er.elementa

import club.sk1er.elementa.constraints.ConstraintType
import club.sk1er.elementa.constraints.resolution.ConstraintVisitor
import club.sk1er.elementa.font.FontProvider
import club.sk1er.mods.core.universal.UGraphics
import club.sk1er.mods.core.universal.UMinecraft.getFontRenderer
import java.awt.Color

class VanillaFontRenderer : FontProvider {
    override var cachedValue: FontProvider = this
    override var recalculate: Boolean = false
    override var constrainTo: UIComponent? = null

    override fun visitImpl(visitor: ConstraintVisitor, type: ConstraintType) {

    }

    override fun getStringWidth(string: String, pointSize: Float): Float {
        return getFontRenderer().getStringWidth(string).toFloat()
    }

    override fun drawString(
        string: String,
        color: Color,
        x: Float,
        y: Float,
        originalPointSize: Float,
        scale: Float,
        shadow: Boolean,
        shadowColor: Color?
    ) {
        UGraphics.scale(scale, scale, 1f)
        if (shadowColor == null) {
            UGraphics.drawString(string, x / scale, y / scale, color.rgb, shadow)
        } else {
            UGraphics.drawString(string, x / scale, y / scale, color.rgb, shadowColor.rgb)
        }
        UGraphics.scale(1 / scale, 1 / scale, 1f)

    }
}